geomstats/geometry/product_manifold.py

```diff
@@ -311,3 +312,263 @@ def is_tangent(self, vector, base_point, atol=gs.atol):
+        """
+        point_ = gs.reshape(base_point, (-1,) + self.base_shape)
+        vector_ = gs.reshape(vector, (-1,) + self.base_shape)
+        each_tangent = self.base_manifold.is_tangent(vector_, point_)
+        reshaped = gs.reshape(each_tangent, (-1, self.n_copies))
+        return gs.all(reshaped, axis=1)
+
+    def to_tangent(self, vector, base_point):
+        """Project a vector to a tangent space of the manifold.
+
+        The tangent space of the product manifold is the direct sum of
+        tangent spaces.
+
+        Parameters
+        ----------
+        vector : array-like, shape=[..., n_copies, base_shape]
+            Vector.
+        base_point : array-like, shape=[..., n_copies, base_shape]
+            Point on the manifold.
+
+        Returns
+        -------
+        tangent_vec : array-like, shape=[..., n_copies, base_shape]
+            Tangent vector at base point.
+        """
+        point_ = gs.reshape(base_point, (-1,) + self.base_shape)
+        vector_ = gs.reshape(vector, (-1,) + self.base_shape)
+        each_tangent = self.base_manifold.to_tangent(vector_, point_)
+        reshaped = gs.reshape(each_tangent, (-1, self.n_copies) + self.base_shape)
+        return gs.squeeze(reshaped)
+
+    def random_point(self, n_samples=1, bound=1.0):
+        """Sample in the product space from the uniform distribution.
+
+        Parameters
+        ----------
+        n_samples : int, optional
+            Number of samples.
+        bound : float
+            Bound of the interval in which to sample for non compact manifolds.
+            Optional, default: 1.
+
+        Returns
+        -------
+        samples : array-like, shape=[..., n_copies, base_shape]
+            Points sampled on the product manifold.
+        """
+        sample = self.base_manifold.random_point(n_samples * self.n_copies, bound)
+        reshaped = gs.reshape(sample, (n_samples, self.n_copies) + self.base_shape)
+        return gs.squeeze(reshaped)
+
+    def projection(self, point):
+        if hasattr(self.base_manifold, "projection"):
+            point_ = gs.reshape(point, (-1,) + self.base_shape)
+            projected = self.base_manifold.projection(point_)
+            reshaped = gs.reshape(projected, (-1, self.n_copies) + self.base_shape)
+            return gs.squeeze(reshaped)
+        raise NotImplementedError(
+            "The base manifold does not implement a projection " "method."
+        )
+
+
+class NFoldMetric(RiemannianMetric):
+    """Class for an n-fold product manifold M^n.
+
+    Define a manifold as the product manifold of n copies of a given base manifold M.
+
+    Parameters
+    ----------
+    base_metric : RiemannianMetric
+        Base metric.
+    n_copies : int
+        Number of replication of the base metric.
+    base_shape : tuple of int
+        Shape of one element of the underlying base manifold
+    """
+
+    def __init__(self, base_metric, n_copies, base_shape):
```

Comment:
Should we do:
self.base_shape = base_metric.shape

and have each metric have a shape attribute too?

It would be more symmetric between the Metric and Manifold classes.